hackathon_site/hardware/views.py

```diff
@@ -31,9 +35,38 @@ def get(self, request, *args, **kwargs):
         return self.list(request, *args, **kwargs)
 
 
-class OrderListView(mixins.ListModelMixin, generics.GenericAPIView):
+class OrderViewSet(
+    mixins.ListModelMixin, mixins.CreateModelMixin, viewsets.GenericViewSet
+):
     queryset = Order.objects.all()
-    serializer_class = OrderSerializer
+    serializer_class = OrderListSerializer
+    serializer_action_classes = {
+        "list": OrderListSerializer,
+        "create": OrderCreateSerializer,
+    }
+
+    def get_serializer_class(self):
+        try:
+            return self.serializer_action_classes[self.action]
+        except (KeyError, AttributeError):
+            return super(OrderViewSet, self).get_serializer_class()
 
     def get(self, request, *args, **kwargs):
         return self.list(request, *args, **kwargs)
+
+    @transaction.atomic
+    def create(self, request, *args, **kwargs):
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        new_order, unfulfilled_hardware_requests = serializer.save()
+        if new_order is None:
+            return Response(
+                {"details": "Cannot be fulfilled"},
+                status=status.HTTP_500_INTERNAL_SERVER_ERROR,
+            )
```

Comment:
this wasnt in the ticket
I'm not creating an order object when the entire order cannot be fulfilled, in which case i think we shouldn't return 201

Reply:
I agree we shouldn't return a 201, but it probably shouldn't be a 500 either. Can we get some more details? If it was something the user did wrong, it should be a 400.

Reply:
Yea I wasn't sure about the code. What details do you want to include?

Reply:
Any details you think should be passed to the frontend for the user to fix.

If it's not enough quantity available, that should just go in the error key of the response and keep the status as 201. Anything that's the user's fault that will stop us from fulfilling an order (constraints not me, hardware ID invalid) should throw a 400 (look up DRF `ValidationError`). It's fine if you validate those one at a time and just throw the first error you hit.

Reply:
done